pyanaconda/modules/payloads/source/cdrom/initialization.py

```diff
@@ -41,14 +43,58 @@ def _do_mount(self):
         log.debug("Trying to detect CD-ROM automatically")
 
         device_tree = STORAGE.get_proxy(DEVICE_TREE)
+
+        device_candidates = device_tree.FindOpticalMedia()
+        stage2_device = self._probe_stage2_for_cdrom(device_tree)
+
+        if stage2_device:
+            device_candidates = [stage2_device] + device_candidates
+
+        device_name = self._search_for_optical_media(device_tree, device_candidates)
+
+        if not device_name:
+            raise SourceSetupError("Found no CD-ROM")
+
+        return device_name
+
+    @staticmethod
+    def _probe_stage2_for_cdrom(device_tree):
+        if "repo" in kernel_arguments:
+            return ""
+
+        log.debug("Testing if inst.stage2 is valid CDRom device source")
+
+        stage2_string = kernel_arguments.get("stage2")
+
+        try:
+            source = SourceFactory.parse_repo_cmdline_string(stage2_string)
+        except PayloadSourceTypeUnrecognized:
+            log.warning("Unknown stage2 method: %s", stage2_string)
+            return ""
+
+        # We HDD here because DVD ISO has inst.stage2=hd:LABEL=....
+        # TODO: Let's return back support of inst.cdrom=<device> which should work based on the
+        # documentation and use that as inst.stage2 parameter for Pungi
+        if not source.is_harddrive:
+            log.debug("Stage2 can't be used as source %s", stage2_string)
+            return ""
+
+        # We can ignore source.path here because DVD ISOs are not using that.
+        stage2_device = device_tree.ResolveDevice(source.partition)
+
+        log.debug("Found possible stage2 default installation source %s", stage2_device)
+
+        return stage2_device
+
+    def _search_for_optical_media(self, device_tree, devices_candidates):
```

Comment:
All device candidates are optical media. We are looking for a valid installation media, so I would rename the function to something like that.